lib/utils/Integers.ml

```diff
@@ -0,0 +1,207 @@
+  type res = Single of t | Multiple of { low : t; high : t }
+
+  let two_16 = Int32.(shift_left one 16)
+  let is_uint16 imm = 0l <= imm && imm < two_16
+  let zero = 0l
+
+  let of_int32 i =
+    if is_uint16 i then Single i
+    else
+      let low, high =
+        (Int32.(logand i 0xffffl), Int32.(shift_right_logical i 16))
+      in
+      Multiple { low; high }
+
+  let encode = Fun.id
+end
+
+(** Module type of 24bit signed integers *)
+module type Int24 = sig
+  type t
+
+  val zero : t
+  val of_int : int -> t option
+  val encode : t -> int32
+end
+
+module Int24 = struct
+  type t = int32
+
+  let zero = 0l
+  let of_int i = if is_int24 i then Some (Int32.of_int i) else None
+  let encode = Fun.id
+end
+
+(** Module type of an offset in the program *)
+module type Offset = sig
+  type t
+
+  val of_int : int -> t option
+  (** Build an offset from an integer *)
+
+  val fit_in_int24 : t -> bool
+  val to_int24 : t -> Int24.t option
+  val pp : Format.formatter -> t -> unit
+end
+
+module Offset = struct
+  type t = int
+
+  let min_offset = first_address - last_address
+  let max_offset = last_address - first_address
+  let is_valid i = min_offset <= i && i <= max_offset
+  let of_int i = if is_valid i then Some i else None
+  let fit_in_int24 t = is_int24 t
+  let to_int24 t = Int24.of_int t
+  let pp ppf t = Format.fprintf ppf "%d" t
+end
+
+(** Module type of an immediate value in the program, ie. a 32bit constant. *)
+module type Immediate = sig
+  type t
+
+  val of_int : int -> t option
+  (** Build an immediate from an integer *)
+
+  val to_uint16 : t -> UInt16.res
+  (** Convert it to a UInt16 *)
+
+  val to_word : t -> Word.t
+  (** Convert it to a word *)
+end
+
+module Immediate = struct
+  type t = int
+
+  let of_int i = if is_int32 i || is_uint32 i then Some i else None
+  let to_uint16 i = UInt16.of_int32 (Int32.of_int i) (* TODO : Check This ! *)
+  let to_word i = Int32.of_int i
+end
+
+(** A 32bit address in the program *)
+module type Address = sig
+  type t
+
+  val first : t
+  val last : t
+  val of_int : int -> t option
+  val next : t -> t option
+  val to_uint16 : t -> UInt16.res
+  val fit_in_uint16 : t -> int -> bool
+  val diff : t -> t -> Offset.t
+  val with_offset : t -> Offset.t -> t
+  val well_defined : t -> Offset.t -> bool
+  val pp : Format.formatter -> t -> unit
+  val is_after : t -> t -> bool
+
+  module Set : Set.S with type elt = t
+
+  type set = Set.t
+
+  module Map : Map.S with type key = t
+
+  type 'a map = 'a Map.t
+end
+
+module Address = struct
+  type t = int
+
+  let first = first_address
+  let last = last_address
+  let of_int i = if first_address <= i && i <= last then Some i else None
+  let next t = of_int (t + 1)
+  let to_uint16 i = UInt16.of_int32 (Int32.of_int i)
+
+  let fit_in_uint16 t i =
+    match of_int (t + i) with Some i -> is_uint16 i | None -> false
+
+  let diff a1 a2 = a1 - a2
+
+  (** Note: The normalization of the address is implementations defined.
```

Comment:
implementation defined et non pas implementation**s** (je crois).

Et : NOOONNN, même en assembleur ont a des implementation defined behavior x)

Reply:
Yep y'a une typo (btw il faut que je check l'orthographe des fichiers de code).

Sinon, le commentaire est là pour expliquer qu'on doit faire un choix lorsqu'on calcule une addresse avec un offset qui fait sortir l'addresse de la plage autorisée (0, 2^32-1). Tous les choix sont possibles (certains logiques, d'autres non). Donc, le résultat du décalage d'une adresse par un offset la faisant sortir de la plage autorisée est bien défini dans l'implantation de l'assembleur ! On fait le choix de faire "comme si" les deux bouts se rejoignaient :

- `Adresse(0) + Offset(-1) = Adresse (2^32 - 1)`
- `Adresse(2^32 - 1) + Offset(1) = Adresse (0)`

Reply:
Fixed (by deletion) in bca7d9cce54d4c6c746fa336784868d00579b141.